File: frontend/_lib/services/user_service.ts
"use client";

import { axiosInstance } from "../axiosInstance";
import { authService } from "../auth";
import { CreateUserPayload, UpdateUserPayload, User } from "@/types/User";

export class UserService {
  private apiUrl: string;

  constructor(baseUrl?: string) {
    this.apiUrl = baseUrl || process.env.NEXT_PUBLIC_API_URL || "http://localhost:8080";
  }

  async getAll(params?: { page?: number; limit?: number; role?: string; search?: string }): Promise<{ data: User[], meta: any }> {
    const res = await axiosInstance.get(`/admin/users`, {
      headers: authService.getAuthHeader(),
      params,
    });
    return {
      data: res.data.data || [],
      meta: res.data.meta || {},
    };
  }

  async getById(id: number): Promise<User | null> {
    try {
      const res = await axiosInstance.get(`/admin/users/${id}`, {
        headers: authService.getAuthHeader(),
      });
      return res.data.user;
    } catch (err) {
      console.error("Failed to fetch user:", err);
      return null;
    }
  }

  async create(payload: CreateUserPayload): Promise<User> {
    const res = await axiosInstance.post(`/admin/users`, payload, {
      headers: authService.getAuthHeader(),
    });
    return res.data;
  }

  // Update user
  async update(id: number, payload: UpdateUserPayload): Promise<User> {
    const res = await axiosInstance.put(`/admin/users/${id}`, payload, {
      headers: authService.getAuthHeader(),
    });
    return res.data;
  }

  // Delete user
  async delete(id: number): Promise<boolean> {
    await axiosInstance.delete(`/admin/users/${id}`, {
      headers: authService.getAuthHeader(),
    });
    return true;
  }

  async activate(id: number): Promise<User> {
    return this.update(id, { isActive: true });
  }

  async deactivate(id: number): Promise<User> {
    return this.update(id, { isActive: false });
  }
}

File: backend/resource/services/superadmin/user_service.go
package superadmin

import (
	"backend/resource/models"
	repo "backend/resource/repositories/superadmin"
)

type PaginatedUsersResponse struct {
	Data  []models.User `json:"data"`
	Total int64         `json:"total"`
	Page  int           `json:"page"`
	Limit int           `json:"limit"`
}

type UserService interface {
	GetUsers(page, limit int, search, role, status, sort string) (*PaginatedUsersResponse, error)
	GetUserByID(id uint) (*models.User, error)
	UpdateUserStatus(id uint, isActive bool) error
	UpdateUser(id uint, updates map[string]interface{}) error
}

type userService struct {
	repo repo.UserRepository
}

func NewUserService(repo repo.UserRepository) UserService {
	return &userService{repo}
}

func (s *userService) GetUsers(page, limit int, search, role, status, sort string) (*PaginatedUsersResponse, error) {
	users, total, err := s.repo.GetAll(page, limit, search, role, status, sort)
	if err != nil {
		return nil, err
	}

	return &PaginatedUsersResponse{
		Data:  users,
		Total: total,
		Page:  page,
		Limit: limit,
	}, nil
}

func (s *userService) GetUserByID(id uint) (*models.User, error) {
	return s.repo.GetByID(id)
}

func (s *userService) UpdateUserStatus(id uint, isActive bool) error {
	return s.repo.UpdateStatus(id, isActive)
}

func (s *userService) UpdateUser(id uint, updates map[string]interface{}) error {
	return s.repo.Update(id, updates)
}
